### PR TITLE
Implement graceful termination and group-based app lifetime tracking

### LINKED ISSUE #1997: Processes are terminated with SIGKILL on Linux, making clean termination impossible
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

proc_t uses boost's group termination feature to kill an application's process group. Boost will always terminate Linux processes by sending a SIGKILL (9) to the kernel, a signal which is not forwarded to processes.

This behaviour means that processes are never allowed to perform their shutdown routines when terminated by Sunshine.

### Expected Behavior

I would expect Sunshine to terminate Linux processes with a SIGTERM (15) signal, which would allow processes to terminate gracefully.

Possibly, a SIGKILL (9) could be sent after a predetermined timeout in case the application takes too long to terminate.

### Additional Context

I'm currently working on a small set of utilities which, among other things, allow Sunshine to properly execute Steam games without having to use detached commands. For this, an auxiliary process is used to launch the game and later monitor the process created by Steam. In case the auxiliary process is killed before the game, it performs some clean up tasks and kills the game process.

The workaround for me will be to break this single process into multiple processes and use Sunshine's command preparations or detached commands. This workaround won't be possible for programs not specifically designed for Sunshine.

### Host Operating System

Linux

### Operating System Version

Fedora 39, kernel 6.6.7-200.fc39.x86_64

### Architecture

64 bit

### Sunshine commit or version

29a1b15

### Package

other (self built)

### GPU Type

AMD

### GPU Model

RX 6800 XT

### GPU Driver/Mesa Version

23.3.1

### Capture Method (Linux Only)

KMS

### Config

```Shell
av1_mode = 0
hevc_mode = 0
```

### Apps

```json
{
    "env": {
        "PATH": "$(PATH):$(HOME)\/.local\/bin"
    },
    "apps": [
        {
            "name": "Steam",
            "image-path": "desktop.png",
            "exclude-global-prep-cmd": "false",
            "auto-detach": "true"
        },
        {
            "name": "Cyberpunk 2077",
            "output": "",
            "cmd": "launch-steam-game \"1091500\" \"Cyberpunk2077.exe\"",
            "exclude-global-prep-cmd": "false",
            "elevated": "false",
            "auto-detach": "true",
            "image-path": "\/home\/games\/.config\/sunshine\/covers\/igdb_1877.png"
        },
        {
            "name": "Grand Theft Auto V",
            "output": "",
            "cmd": "launch-steam-game \"271590\" \"PlayGTAV.exe\"",
            "exclude-global-prep-cmd": "false",
            "elevated": "false",
            "auto-detach": "true",
            "image-path": "\/home\/games\/.config\/sunshine\/covers\/igdb_1020.png"
        }
    ]
}
```

### Relevant log output

```shell
[2024:01:07:12:53:24]: Info: // Testing for available encoders, this may generate errors. You can safely ignore those errors. //
[2024:01:07:12:53:24]: Info: Trying encoder [nvenc]
[2024:01:07:12:53:24]: Info: Screencasting with KMS
[2024:01:07:12:53:24]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:24]: Info: Screencasting with KMS
[2024:01:07:12:53:24]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:24]: Info: SDR color coding [Rec. 601]
[2024:01:07:12:53:24]: Info: Color depth: 8-bit
[2024:01:07:12:53:24]: Info: Color range: [JPEG]
[2024:01:07:12:53:24]: Error: [AVHWDeviceContext @ 0x7f84d415d3c0] Cannot load libcuda.so.1
[2024:01:07:12:53:24]: Error: [AVHWDeviceContext @ 0x7f84d415d3c0] Could not dynamically load CUDA
[2024:01:07:12:53:24]: Error: Failed to create a CUDA device: Operação não permitida
[2024:01:07:12:53:24]: Info: Screencasting with KMS
[2024:01:07:12:53:24]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:24]: Info: SDR color coding [Rec. 601]
[2024:01:07:12:53:24]: Info: Color depth: 8-bit
[2024:01:07:12:53:24]: Info: Color range: [JPEG]
[2024:01:07:12:53:24]: Error: [AVHWDeviceContext @ 0x7f84d40ed3c0] Cannot load libcuda.so.1
[2024:01:07:12:53:24]: Error: [AVHWDeviceContext @ 0x7f84d40ed3c0] Could not dynamically load CUDA
[2024:01:07:12:53:24]: Error: Failed to create a CUDA device: Operação não permitida
[2024:01:07:12:53:24]: Info: Encoder [nvenc] failed
[2024:01:07:12:53:24]: Info: Trying encoder [vaapi]
[2024:01:07:12:53:24]: Info: Screencasting with KMS
[2024:01:07:12:53:24]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:24]: Info: Screencasting with KMS
[2024:01:07:12:53:24]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:24]: Info: SDR color coding [Rec. 601]
[2024:01:07:12:53:24]: Info: Color depth: 8-bit
[2024:01:07:12:53:24]: Info: Color range: [JPEG]
[2024:01:07:12:53:24]: Warning: [h264_vaapi @ 0x7f84d437fec0] Driver does not support some wanted packed headers (wanted 0xd, found 0x1).
[2024:01:07:12:53:24]: Info: Screencasting with KMS
[2024:01:07:12:53:24]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:24]: Info: SDR color coding [Rec. 601]
[2024:01:07:12:53:24]: Info: Color depth: 8-bit
[2024:01:07:12:53:24]: Info: Color range: [JPEG]
[2024:01:07:12:53:24]: Warning: [hevc_vaapi @ 0x7f84d48fdc80] Driver does not support some wanted packed headers (wanted 0xd, found 0x1).
[2024:01:07:12:53:24]: Info: Screencasting with KMS
[2024:01:07:12:53:24]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:24]: Info: SDR color coding [Rec. 601]
[2024:01:07:12:53:24]: Info: Color depth: 8-bit
[2024:01:07:12:53:24]: Info: Color range: [JPEG]
[2024:01:07:12:53:24]: Error: [av1_vaapi @ 0x7f84d48fdc80] No usable encoding entrypoint found for profile VAProfileAV1Profile0 (32).
[2024:01:07:12:53:24]: Error: Could not open codec [av1_vaapi]: Função não implementada
[2024:01:07:12:53:24]: Info: Screencasting with KMS
[2024:01:07:12:53:24]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:24]: Info: SDR color coding [Rec. 601]
[2024:01:07:12:53:24]: Info: Color depth: 8-bit
[2024:01:07:12:53:24]: Info: Color range: [JPEG]
[2024:01:07:12:53:24]: Error: [av1_vaapi @ 0x7f84d491ff40] No usable encoding entrypoint found for profile VAProfileAV1Profile0 (32).
[2024:01:07:12:53:24]: Error: Could not open codec [av1_vaapi]: Função não implementada
[2024:01:07:12:53:24]: Info: Screencasting with KMS
[2024:01:07:12:53:24]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:24]: Info: SDR color coding [Rec. 601]
[2024:01:07:12:53:24]: Info: Color depth: 8-bit
[2024:01:07:12:53:24]: Info: Color range: [JPEG]
[2024:01:07:12:53:24]: Error: [av1_vaapi @ 0x7f84d437fec0] No usable encoding entrypoint found for profile VAProfileAV1Profile0 (32).
[2024:01:07:12:53:24]: Error: Could not open codec [av1_vaapi]: Função não implementada
[2024:01:07:12:53:25]: Info: Screencasting with KMS
[2024:01:07:12:53:25]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:25]: Info: SDR color coding [Rec. 601]
[2024:01:07:12:53:25]: Info: Color depth: 8-bit
[2024:01:07:12:53:25]: Info: Color range: [JPEG]
[2024:01:07:12:53:25]: Error: [av1_vaapi @ 0x7f84d420f3c0] No usable encoding entrypoint found for profile VAProfileAV1Profile0 (32).
[2024:01:07:12:53:25]: Error: Could not open codec [av1_vaapi]: Função não implementada
[2024:01:07:12:53:25]: Info: Screencasting with KMS
[2024:01:07:12:53:25]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:25]: Info: SDR color coding [Rec. 709]
[2024:01:07:12:53:25]: Info: Color depth: 10-bit
[2024:01:07:12:53:25]: Info: Color range: [JPEG]
[2024:01:07:12:53:25]: Error: [AVHWFramesContext @ 0x7f84d430bfc0] Unsupported format: yuv420p10le.
[2024:01:07:12:53:25]: Info: 
[2024:01:07:12:53:25]: Info: // Ignore any errors mentioned above, they are not relevant. //
[2024:01:07:12:53:25]: Info: 
[2024:01:07:12:53:25]: Info: Found H.264 encoder: h264_vaapi [vaapi]
[2024:01:07:12:53:25]: Info: Found HEVC encoder: hevc_vaapi [vaapi]
[2024:01:07:12:53:25]: Info: Executing: [launch-steam-game "1091500" "Cyberpunk2077.exe"] in ["/home/games/.local/bin"]
[2024:01:07:12:53:25]: Info: CLIENT CONNECTED
[2024:01:07:12:53:25]: Info: Screencasting with KMS
[2024:01:07:12:53:25]: Info: Found monitor for DRM screencasting
[2024:01:07:12:53:25]: Info: SDR color coding [Rec. 601]
[2024:01:07:12:53:25]: Info: Color depth: 8-bit
[2024:01:07:12:53:25]: Info: Color range: [MPEG]
[2024:01:07:12:53:25]: Warning: [hevc_vaapi @ 0x7f847421f380] Driver does not support some wanted packed headers (wanted 0xd, found 0x1).
[2024:01:07:12:53:25]: Info: Setting default sink to: [sink-sunshine-stereo]
[2024:01:07:12:53:25]: Info: Found default monitor by name: sink-sunshine-stereo.monitor
[2024:01:07:12:54:05]: Info: CLIENT DISCONNECTED
[2024:01:07:12:54:05]: Info: Setting default sink to: [alsa_output.pci-0000_2b_00.1.hdmi-stereo-extra3]
```

### PR DESCRIPTION
**Pull request by cgutman**
## Description
This PR introduces OS-specific graceful termination logic to allow apps to terminate normally rather than be abruptly killed when the client requests them to be stopped. On Windows, we send `WM_CLOSE` to all windows that are owned by processes in the app's process group. On Mac and Linux, we send `SIGTERM` to the process group.

This PR also introduces group-based app lifetime tracking which uses the process group to determine if the "app" is still running rather than just looking at the status of the first process we launched. This group-based tracking takes precedence over auto-detach if both are enabled and the initial process terminates after launching a child, since group-based tracking still allows full functionality of the running/not-running state and the Quit App button in Moonlight, unlike auto-detach. This feature should largely eliminate the need for using detached commands for "normal" apps.

I'm still keeping auto-detach around because there are some cases where a process may spawn a child into another process group or indirectly via another running process, which will still fool our group-based process tracking.

This PR also contains a fix to address a regression in #1974 that would cause abnormal termination of Sunshine if a process exited before the user quit it in Moonlight.

### Screenshot
image

### Issues Fixed or Closed
Fixes #1997

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

File: src/process.h [hunk 1]
```diff
@@ -53,6 +53,8 @@ namespace proc {
     std::string id;
     bool elevated;
     bool auto_detach;
+    bool wait_all;
+    std::chrono::seconds exit_timeout;
```
Comment (ReenigneArcher) [inline src/process.h:57]: Should the comment above this be updated?

`That is why some applications need be run and forgotten about`

File: src/process.cpp [hunk 2]
```diff
@@ -625,6 +664,8 @@ namespace proc {
 
         ctx.elevated = elevated.value_or(false);
         ctx.auto_detach = auto_detach.value_or(true);
+        ctx.wait_all = wait_all.value_or(true);
+        ctx.exit_timeout = std::chrono::seconds { exit_timeout.value_or(5) };
```
Comment (ReenigneArcher) [inline src/process.cpp:668]: This section sets the value to our default for apps which haven't been edited yet?

**Review by ReenigneArcher — DISMISSED** [review]
Another nice enhancement!

File: src/process.cpp (hunk 2 above)
Comment (cgutman) [reply src/process.cpp:668]: Yep, this is basically the equivalent of our `config_t` defaults in config.cpp.

File: src/process.h (hunk 1 above)
Comment (cgutman) [reply src/process.h:63]: Yep, updated it.

**Review by ReenigneArcher — APPROVED** [review]

**Pull request merged**